Assets/Scripts/Character/CharacterController.cs

```diff
@@ -0,0 +1,121 @@
+    }
+
+    /*
+     * Usually the physics engine handles friction/drag automatically. 
+     * But I'm too lazy to assign friction to every material and drag shouldn't be used,
+     * so here is this function that gets called each timestep and applies friction automatically.
+     */
+    private Vector3 AddFriction(Vector3 currentVel, float friction, bool isYAffected = false) {
+        float speed = currentVel.magnitude;
+        Vector3 newVel = Vector3.zero;
+
+        // We need to slow down the speed if it is above zero
+        if (speed != 0) {
+            float drop = friction * speed;
+            float scale = Mathf.Max(0, speed - drop) / speed;
+
+            newVel = Vector3.Scale(currentVel, new Vector3(scale, isYAffected ? scale : 1, scale));
+        }
+
+        return newVel;
+    }
+
+    void FixedUpdate() {
+        bool isGrounded = IsGrounded();
+        float FRICTION = isGrounded ? GROUND_FRICTION : AIR_FRICTION;
+        float ACCELERATION = (isGrounded ? GROUND_ACCELERATION : AIR_ACCELERATION) * sprintMultiplier;
+        float MAX_VELOCITY = (isGrounded ? GROUND_MAX_VELOCITY : AIR_MAX_VELOCITY) * sprintMultiplier;
+        Vector3 currentVelocity = _rigidbody.velocity;
+
+        Vector3 deltaVelocity = moveDirection * ACCELERATION;
+        Vector3 newVelocity = new Vector3(currentVelocity.x + deltaVelocity.x, 0, currentVelocity.z + deltaVelocity.z);
+        Vector3 newVelocityClamped = Vector3.ClampMagnitude(newVelocity, MAX_VELOCITY);
+
+        Vector3 finalVelocity = newVelocityClamped + new Vector3(0, currentVelocity.y, 0);
+        Vector3 finalVelocityWithFriction = AddFriction(finalVelocity, FRICTION);
+        _rigidbody.velocity = finalVelocityWithFriction;
+        velocity = _rigidbody.velocity;
+    }
+
+    // Helper function to convert a wrapped angle to a non wrapped angle (etc. 270 -> -90)
+    private Vector3 ConvertAngle(Vector3 transformAngle) {
+        float x = (transformAngle.x > 180) ? transformAngle.x - 360 : transformAngle.x;
+        float y = (transformAngle.y > 180) ? transformAngle.y - 360 : transformAngle.y;
+        float z = (transformAngle.z > 180) ? transformAngle.z - 360 : transformAngle.z;
+
+        return new Vector3(x, y, z);
+    }
+
+    // Function to set the camera perspective directly
+    public void SetView(float pitch, float yaw) {
+        Vector3 cameraRotation = ConvertAngle(_cameraTransform.localRotation.eulerAngles);
+        pitchDegree = Mathf.Clamp(pitch, -MAX_PITCH_DEGREE, MAX_PITCH_DEGREE);
+        _cameraTransform.localRotation = Quaternion.Euler(pitchDegree, cameraRotation.y, cameraRotation.z);
+
+        Vector3 playerRotation = ConvertAngle(_bodyTransform.rotation.eulerAngles);
+        yawDegree = yaw;
+        _bodyTransform.rotation = Quaternion.Euler(playerRotation.x, yawDegree, playerRotation.z);
+    }
+
+    public bool IsGrounded() {
+        return Physics.CheckSphere(_groundCheckTransform.position, 1f, _groundLayer);
```

Comment:
What may be better here instead of using a transform for the ground check position, is to instead reference the collider's bottom-most point. So let's say we have _movementCollider, we could do groundCheckPosition = new Vector3(_movementCollider.bounds.center.x, _movementCollider.bounds.min.y, _movementCollider.bounds.center.z).
What you have now works fine though, so this is definitely an optional suggestion.

Reply:
Implemented